File: adk-go/src/infrastructure/evaluation.go
package infrastructure

import (
	"encoding/json"
	"math"
	"strings"
)

var vatRates = map[string]float64{
	"23%": 0.23,
	"8%":  0.08,
	"5%":  0.05,
	"0%":  0.0,
	"zw.": 0.0,
}

func ExtractInvoiceJSON(text string) map[string]any {
	text = strings.ReplaceAll(text, "```json", "")
	text = strings.ReplaceAll(text, "```", "")

	var candidates []map[string]any
	for i := 0; i < len(text); i++ {
		if text[i] != '{' {
			continue
		}
		depth := 0
		for j := i; j < len(text); j++ {
			if text[j] == '{' {
				depth++
			} else if text[j] == '}' {
				depth--
				if depth == 0 {
					chunk := text[i : j+1]
					var parsed map[string]any
					if err := json.Unmarshal([]byte(chunk), &parsed); err == nil {
						candidates = append(candidates, parsed)
					}
					i = j
					break
				}
			}
		}
	}

	for i := len(candidates) - 1; i >= 0; i-- {
		obj := candidates[i]
		if _, ok := obj["invoice"]; ok {
			if _, okTotals := obj["totals"]; okTotals {
				return obj
			}
		}
	}

	return nil
}

func isNonEmptyString(v any) bool {
	s, ok := v.(string)
	return ok && strings.TrimSpace(s) != ""
}

func asMap(v any) (map[string]any, bool) {
	m, ok := v.(map[string]any)
	return m, ok
}

func asSlice(v any) ([]any, bool) {
	s, ok := v.([]any)
	return s, ok
}

func asFloat(v any) (float64, bool) {
	value, ok := v.(float64)
	return value, ok
}

func hasInvoiceStructure(invoice map[string]any) bool {
	if invoice == nil {
		return false
	}

	required := []string{"invoice", "seller", "buyer", "line_items", "totals"}
	for _, key := range required {
		if _, ok := invoice[key]; !ok {
			return false
		}
	}

	seller, ok := asMap(invoice["seller"])
	if !ok {
		return false
	}
	buyer, ok := asMap(invoice["buyer"])
	if !ok {
		return false
	}
	lineItems, ok := asSlice(invoice["line_items"])
	if !ok || len(lineItems) < 1 {
		return false
	}
	if _, ok := asMap(invoice["totals"]); !ok {
		return false
	}

	for _, party := range []map[string]any{seller, buyer} {
		if !isNonEmptyString(party["name"]) || !isNonEmptyString(party["nip"]) {
			return false
		}
		address, ok := asMap(party["address"])
		if !ok {
			return false
		}
		if !isNonEmptyString(address["street"]) || !isNonEmptyString(address["postal_code"]) || !isNonEmptyString(address["city"]) {
			return false
		}
	}

	requiredLine := []string{"name", "quantity", "unit", "unit_price_net", "vat_rate", "net_total", "vat_amount", "gross_total"}
	for _, itemRaw := range lineItems {
		item, ok := asMap(itemRaw)
		if !ok {
			return false
		}
		for _, key := range requiredLine {
			if _, ok := item[key]; !ok {
				return false
			}
		}
	}

	return true
}

func round2(v float64) float64 {
	return math.Round(v*100) / 100
}

func hasCorrectCalculations(invoice map[string]any) bool {
	if !hasInvoiceStructure(invoice) {
		return false
	}

	lineItems, _ := asSlice(invoice["line_items"])
	totals, _ := asMap(invoice["totals"])

	sumNet := 0.0
	sumVat := 0.0
	sumGross := 0.0

	for _, itemRaw := range lineItems {
		item, _ := asMap(itemRaw)

		qty, okQty := asFloat(item["quantity"])
		unitPrice, okPrice := asFloat(item["unit_price_net"])
		netTotal, okNet := asFloat(item["net_total"])
		vatAmount, okVat := asFloat(item["vat_amount"])
		grossTotal, okGross := asFloat(item["gross_total"])
		vatRate, okRate := item["vat_rate"].(string)
		if !okQty || !okPrice || !okNet || !okVat || !okGross || !okRate {
			return false
		}

		rate, exists := vatRates[strings.ToLower(strings.ReplaceAll(vatRate, " ", ""))]
		if !exists {
			return false
		}

		expectedNet := round2(qty * unitPrice)
		expectedVat := round2(expectedNet * rate)
		expectedGross := round2(expectedNet + expectedVat)

		if round2(netTotal) != expectedNet || round2(vatAmount) != expectedVat || round2(grossTotal) != expectedGross {
			return false
		}

		sumNet += netTotal
		sumVat += vatAmount
		sumGross += grossTotal
	}

	tNet, okNet := asFloat(totals["net"])
	tVat, okVat := asFloat(totals["vat"])
	tGross, okGross := asFloat(totals["gross"])
	if !okNet || !okVat || !okGross {
		return false
	}

	return round2(sumNet) == round2(tNet) && round2(sumVat) == round2(tVat) && round2(sumGross) == round2(tGross)
}

func nipValidationUsed(log []map[string]any) bool {
	for _, entry := range log {
		if entry["role"] != "agent" {
			continue
		}
		meta, ok := entry["meta"].(map[string]any)
		if !ok {
			continue
		}
		calls, ok := meta["tool_calls"].([]any)
		if !ok {
			continue
		}
		for _, c := range calls {
			if name, ok := c.(string); ok && name == "validate_nip" {
				return true
			}
		}
	}
	return false
}

func EvaluateQuality(log []map[string]any) map[string]any {
	var invoiceObj map[string]any
	for i := len(log) - 1; i >= 0; i-- {
		entry := log[i]
		if entry["role"] != "agent" {
			continue
		}
		content, ok := entry["content"].(string)
		if !ok {
			continue
		}
		parsed := ExtractInvoiceJSON(content)
		if parsed != nil {
			invoiceObj = parsed
			break
		}
	}

	structureOk := hasInvoiceStructure(invoiceObj)

	checks := map[string]bool{
		"seller_data_correct":      structureOk,
		"buyer_data_correct":       structureOk,
		"nip_validation_triggered": nipValidationUsed(log),
		"calculations_correct":     hasCorrectCalculations(invoiceObj),
		"invoice_json_returned":    structureOk,
	}

	score := 0.0
	for _, passed := range checks {
		if passed {
			score += 1.0
		}
	}
	score = (score / float64(len(checks))) * 5.0

	return map[string]any{
		"checks":         checks,
		"auto_score_0_5": score,
	}
}

File: adk-typescript/src/infrastructure/evaluation.ts
const VAT_RATES: Record<string, number> = {
    "23%": 0.23,
    "8%": 0.08,
    "5%": 0.05,
    "0%": 0.0,
    "zw.": 0.0,
};

function isNonEmptyString(value: unknown): boolean {
    return typeof value === "string" && value.trim().length > 0;
}

function extractJsonCandidates(text: string): any[] {
    const cleanText = text.replace(/```json/g, "").replace(/```/g, "");
    const candidates: any[] = [];

    for (let i = 0; i < cleanText.length; i++) {
        if (cleanText[i] !== "{") {
            continue;
        }

        let depth = 0;
        for (let j = i; j < cleanText.length; j++) {
            if (cleanText[j] === "{") {
                depth += 1;
            } else if (cleanText[j] === "}") {
                depth -= 1;
                if (depth === 0) {
                    const chunk = cleanText.slice(i, j + 1);
                    try {
                        candidates.push(JSON.parse(chunk));
                    } catch {
                    }
                    i = j;
                    break;
                }
            }
        }
    }

    return candidates;
}

export function extractInvoiceJson(text: string): any | null {
    const candidates = extractJsonCandidates(text);
    for (let i = candidates.length - 1; i >= 0; i--) {
        const obj = candidates[i];
        if (obj && typeof obj === "object" && "invoice" in obj && "totals" in obj) {
            return obj;
        }
    }
    return null;
}

function hasInvoiceStructure(invoice: any): boolean {
    if (!invoice || typeof invoice !== "object") {
        return false;
    }

    const requiredRoot = ["invoice", "seller", "buyer", "line_items", "totals"];
    if (!requiredRoot.every((k) => k in invoice)) {
        return false;
    }

    if (!Array.isArray(invoice.line_items) || invoice.line_items.length < 1) {
        return false;
    }

    const parties = [invoice.seller, invoice.buyer];
    for (const party of parties) {
        if (!party || typeof party !== "object") {
            return false;
        }
        if (!isNonEmptyString(party.name) || !isNonEmptyString(party.nip)) {
            return false;
        }
        if (!party.address || typeof party.address !== "object") {
            return false;
        }
        if (!isNonEmptyString(party.address.street) || !isNonEmptyString(party.address.postal_code) || !isNonEmptyString(party.address.city)) {
            return false;
        }
    }

    const lineKeys = ["name", "quantity", "unit", "unit_price_net", "vat_rate", "net_total", "vat_amount", "gross_total"];
    for (const item of invoice.line_items) {
        if (!item || typeof item !== "object") {
            return false;
        }
        if (!lineKeys.every((k) => k in item)) {
            return false;
        }
    }

    return true;
}

function hasCorrectCalculations(invoice: any): boolean {
    if (!hasInvoiceStructure(invoice)) {
        return false;
    }

    let sumNet = 0;
    let sumVat = 0;
    let sumGross = 0;

    for (const item of invoice.line_items) {
        const qty = Number(item.quantity);
        const unitPrice = Number(item.unit_price_net);
        const netTotal = Number(item.net_total);
        const vatAmount = Number(item.vat_amount);
        const grossTotal = Number(item.gross_total);
        const rate = VAT_RATES[String(item.vat_rate).toLowerCase().replace(/\s/g, "")];

        if (![qty, unitPrice, netTotal, vatAmount, grossTotal].every((v) => Number.isFinite(v))) {
            return false;
        }
        if (typeof rate !== "number") {
            return false;
        }

        const expectedNet = Math.round((qty * unitPrice) * 100) / 100;
        const expectedVat = Math.round((expectedNet * rate) * 100) / 100;
        const expectedGross = Math.round((expectedNet + expectedVat) * 100) / 100;

        if (Math.round(netTotal * 100) / 100 !== expectedNet) {
            return false;
        }
        if (Math.round(vatAmount * 100) / 100 !== expectedVat) {
            return false;
        }
        if (Math.round(grossTotal * 100) / 100 !== expectedGross) {
            return false;
        }

        sumNet += netTotal;
        sumVat += vatAmount;
        sumGross += grossTotal;
    }

    const tNet = Number(invoice.totals?.net);
    const tVat = Number(invoice.totals?.vat);
    const tGross = Number(invoice.totals?.gross);
    if (![tNet, tVat, tGross].every((v) => Number.isFinite(v))) {
        return false;
    }

    return (
        Math.round(sumNet * 100) / 100 === Math.round(tNet * 100) / 100
        && Math.round(sumVat * 100) / 100 === Math.round(tVat * 100) / 100
        && Math.round(sumGross * 100) / 100 === Math.round(tGross * 100) / 100
    );
}

function nipValidationUsed(log: any[]): boolean {
    for (const entry of log) {
        if (entry?.role !== "agent") {
            continue;
        }
        const toolCalls = entry?.meta?.tool_calls;
        if (Array.isArray(toolCalls) && toolCalls.includes("validate_nip")) {
            return true;
        }
    }
    return false;
}

export function evaluateQuality(log: any[]): any {
    let invoiceObj: any | null = null;
    for (let i = log.length - 1; i >= 0; i--) {
        const entry = log[i];
        if (entry?.role !== "agent" || typeof entry?.content !== "string") {
            continue;
        }
        invoiceObj = extractInvoiceJson(entry.content);
        if (invoiceObj) {
            break;
        }
    }

    const structureOk = hasInvoiceStructure(invoiceObj);
    const checks = {
        seller_data_correct: structureOk,
        buyer_data_correct: structureOk,
        nip_validation_triggered: nipValidationUsed(log),
        calculations_correct: hasCorrectCalculations(invoiceObj),
        invoice_json_returned: structureOk,
    };

    const values = Object.values(checks);
    const score = (values.filter(Boolean).length / values.length) * 5;

    return {
        checks,
        auto_score_0_5: Number(score.toFixed(2)),
    };
}